### PR TITLE
add error message if a mutable field is initialized with a reference type

### LINKED ISSUE #6: References to mutable data in static initializers potentially error prone
first you should understand why it does this - all initializers go through the ctfe engine then copy the result into the initializer. when it is a value type, copying this gives pretty unsurprising results

but when it is a reference type, copying the reference is only one layer deep. the thing pointed to by the reference is still the single instance

i would say it is a frequently asked question, this has come up m any times. if you come from Java and do like

Class A {
    Object o = new Whatever;
}

this gives very different results in D. again this behavior can be very useful if you know exactly what it is, but odds are you don't.

similarly

struct S {
    int[] a = [1,2,3];
}

perhaps it should just be like 

error: mutable reference type assigned to initial value. move to constructor if you want a unique instance per instance, make immutable if you want a shared unchanging thing, or mark with @core.attribute.ItsOkIKnowWhatItDoesAndWantThat if it is ok, you know what it does, and you want that

### PR DESCRIPTION
**Pull request by brianush1**
would fix #6

### CODE REVIEW

**Comment by brianush1** [comment]
ugh it didn't commit my changes to core.attribute nvm don't merge this yet

**Comment by brianush1** [comment]
alright should be good now

File: compiler/src/dmd/initsem.d [hunk 1]
```diff
@@ -110,6 +110,49 @@ extern(C++) Initializer initializerSemantic(Initializer init, Scope* sc, ref Typ
...
+                return 0;
+            });
+
+            if (badType)
+            {
+                error(init.loc, "mutable reference type assigned in initializer");
+                errorSupplemental(init.loc, "- initialize the field in a constructor if you want a unique value per instance,");
+                errorSupplemental(init.loc, "- mark the field as `immutable` or `const` if you want a shared, unchanging reference,");
+                errorSupplemental(init.loc, "- or mark the field with `@core.attribute.singletonInitializer` to silence this error");
+                return false;
```
Comment (maxhaton) [inline compiler/src/dmd/initsem.d:150]:
```
if (bad) {
  // do the thing
  return
}
```
prefer
```
if (good) {
  return
}
// do the thing.
```

File: druntime/src/core/attribute.d [hunk 2]
```diff
@@ -302,3 +302,8 @@ enum mustuse;
  * This is only allowed on `shared` static constructors, not thread-local module constructors.
  */
 enum standalone;
+
+/** 
+ * Use this attribute to indicate that you wish to initialize a field with a mutable reference type.
+ */
+enum singletonInitializer;
```
Comment (maxhaton) [inline druntime/src/core/attribute.d:309]: > singletonInitializer

Where does the name come from?

**Comment by maxhaton** [comment]
This should be one commit

File: compiler/src/dmd/initsem.d (hunk 1 above)
Comment (maxhaton) [reply compiler/src/dmd/initsem.d]: good

File: druntime/src/core/attribute.d (hunk 2 above)
Comment (brianush1) [reply druntime/src/core/attribute.d:309]: The reference you initialize the variable with is a singleton, there is only one instance of it, which is created at compile-time. Do you have a better name? I'm open to ideas

Comment (maxhaton) [reply druntime/src/core/attribute.d:309]: `mutableInit`?

**Comment by adamdruppe** [comment]
I'm actually surprised how many things this is reporting in dmd and phobos.... some of them are functions, we might want to relax it on that too because functions point to immutable data (the code segment) too.

**Comment by brianush1** [comment]
I tried it on phobos and it compiled without complaints... weird

**Comment by brianush1** [comment]
> I tried it on phobos and it compiled without complaints... weird

nevermind I'm silly, I confused phobos with druntime

**Comment by adamdruppe** [comment]
I got it working as part of my monster megacommit, the monorepo has the fixes. but we can prolly make it work ebtter in some cases like if you're assigning a struct of just strings and function pointers that's basically still a value type. still meh it is fine

i kinda worry about pulling changes out of that huge monorepo thing. there were big commits and small commits mixed together and i prolly broke stuff

but meh we can fix it as it happens from here on out

**Comment by adamdruppe** [comment]
Windows' HANDLE is another one that probably shouldn't trigger it but does.

**Comment by adamdruppe** [comment]
I'm gonna hack that out by making `void*` an exception to the rule too.

**Pull request merged**